src/oid4vp/holder.rs

```diff
@@ -0,0 +1,511 @@
+                None => vec![],
+                Some(vdc_collection) => {
+                    let credential_ids = vdc_collection.all_entries().await?;
+
+                    futures_util::stream::iter(credential_ids)
+                        .filter_map(|id| async move { vdc_collection.get(id).await.ok().flatten() })
+                        .collect::<Vec<Arc<Credential>>>()
+                        .await
+                }
+            },
+        }
+        .into_iter()
+        .filter_map(|cred| cred.try_into_parsed().ok())
+        .filter_map(
+            |cred| match cred.check_presentation_definition(definition) {
+                true => Some(cred),
+                false => None,
+            },
+        )
+        .collect::<Vec<Arc<ParsedCredential>>>();
+
+        Ok(credentials)
+    }
+
+    // Construct a DescriptorMap for the presentation submission based on the
+    // credentials returned from the VDC collection.
+    fn create_descriptor_map(
+        &self,
+        credential: &Arc<ParsedCredential>,
+        input_descriptor_id: String,
+    ) -> Result<DescriptorMap, OID4VPError> {
+        Ok(DescriptorMap::new(
+            input_descriptor_id,
+            ClaimFormatDesignation::JwtVpJson,
+            "$".into(),
+        )
+        // Credentials will be nested within a `vp` JSON object.
+        .set_path_nested(DescriptorMap::new(
+            credential.id().to_string(),
+            ClaimFormatDesignation::from(credential.format()),
+            "$.verifiableCredential".into(),
+        )))
+    }
+
+    // Internal method for returning the `PermissionRequest` for an oid4vp request.
+    async fn permission_request(
+        &self,
+        request: &AuthorizationRequestObject,
+    ) -> Result<Arc<PermissionRequest>, OID4VPError> {
+        // Resolve the presentation definition.
+        let presentation_definition = request
+            .resolve_presentation_definition(self.http_client())
+            .await
+            .map_err(|e| OID4VPError::PresentationDefinitionResolution(e.to_string()))?
+            .into_parsed();
+
+        let credentials = self
+            .search_credentials_vs_presentation_definition(&presentation_definition)
+            .await?;
+
+        Ok(PermissionRequest::new(
+            presentation_definition.clone(),
+            credentials.clone(),
+        ))
+    }
+
+    // Internal method for creating a verifiable presentation.
+    async fn create_unencoded_verifiable_presentation(
+        &self,
+        request: &AuthorizationRequestObject,
+        credential: &Arc<ParsedCredential>,
+    ) -> Result<VpToken, OID4VPError> {
+        match request.client_id_scheme() {
+            ClientIdScheme::Did => self.verifiable_presentation_did(request, credential).await,
+            _ => Err(OID4VPError::InvalidClientIdScheme(
+                "Only DID client ID scheme is supported.".to_string(),
+            )),
+        }
```

Comment:
The kind of verifiable presentation that we create depends on the credential type and the presentation definition, not the client_id_scheme.